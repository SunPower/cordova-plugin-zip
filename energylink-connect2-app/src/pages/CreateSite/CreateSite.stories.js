import React from 'react'
import { storiesOf } from '@storybook/react'
import CreateSite from '.'

storiesOf('CreateSite Page', module).add('Simple', () => (
  <div className="full-min-height pl-10 pr-10">
    <CreateSite />
  </div>
))
