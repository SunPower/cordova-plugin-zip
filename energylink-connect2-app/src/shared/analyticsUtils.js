import {
  __,
  compose,
  curry,
  gt,
  isEmpty,
  join,
  lensIndex,
  map,
  not,
  over,
  props,
  propSatisfies,
  toLower,
  toUpper,
  when
} from 'ramda'

export const isNotEmpty = compose(not, isEmpty)

export const capitalizeWord = when(
  propSatisfies(gt(__, 0), 'length'),
  compose(join(''), over(lensIndex(0), toUpper), toLower)
)

const stringValueOrEmptyString = stringValue => stringValue || ''

const getStringProps = curry(compose(map(stringValueOrEmptyString), props))

export const getUserProfile = getStringProps([
  'uniqueId',
  'firstName',
  'lastName',
  'email',
  'dealerName',
  'recordType'
])
