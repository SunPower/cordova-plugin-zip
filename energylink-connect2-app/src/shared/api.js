import Swagger from 'swagger-client'
let apiParty, apiSite, apiFirmware, apiScanSN

const requestOptions = access_token => ({
  requestInterceptor: req => {
    if (access_token) req.headers['Authorization'] = `Bearer ${access_token}`
    return req
  }
})

export async function getApiFirmware() {
  if (!apiFirmware)
    apiFirmware = await Swagger(process.env.REACT_APP_SWAGGER_FIRMWARE)
  return apiFirmware
}
export async function getApiPVS() {
  return await Swagger(process.env.REACT_APP_PVS_ADDRESS)
}
export async function getApiDevice(access_token) {
  return await Swagger(
    process.env.REACT_APP_SWAGGER_DEVICE,
    requestOptions(access_token)
  )
}
export async function getApiParty(access_token) {
  apiParty = await Swagger(
    process.env.REACT_APP_SWAGGER_PARTY,
    requestOptions(access_token)
  )
  return apiParty
}
export async function getApiSearch(access_token) {
  return await Swagger(
    process.env.REACT_APP_SWAGGER_SEARCH,
    requestOptions(access_token)
  )
}
export async function getApiSite(access_token) {
  if (!apiSite)
    apiSite = await Swagger(
      process.env.REACT_APP_SWAGGER_SITE,
      requestOptions(access_token)
    )
  return apiSite
}
export async function getApiScanSN() {
  if (!apiScanSN)
    apiScanSN = await Swagger(process.env.REACT_APP_SCAN_SERIAL_NUMBERS)
  return apiScanSN
}
export async function getApiAuth(access_token) {
  return await Swagger(
    process.env.REACT_APP_SWAGGER_AUTH,
    requestOptions(access_token)
  )
}
export const storageSwaggerTag = 'Commissioning'
