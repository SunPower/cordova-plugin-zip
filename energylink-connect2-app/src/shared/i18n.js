import { useMemo } from 'react'
import { useSelector } from 'react-redux'

import * as languages from './locales'
import * as localizedComponents from './localizedComponents'

const getTranslate = translations => (key, ...params) => {
  let value = translations[key] || key
  if (params) {
    params.forEach((p, idx) => {
      value = value.replace(`{${idx}}`, p)
    })
  }
  return value
}

export const useI18n = () => {
  const language = useSelector(state => state.language)
  const { locale = 'en' } = language
  return useMemo(() => getTranslate(languages[locale]), [locale])
}

export const useI18nComponent = () => {
  const language = useSelector(state => state.language)
  const { locale = 'en' } = language || {}

  return useMemo(
    () => componentName => {
      const path = `${locale}_${componentName}`
      const Component = localizedComponents[path] || {}
      if (typeof Component !== 'function') {
        throw new Error(`Component not found: ${path}`)
      }
      return Component()
    },
    [locale]
  )
}
