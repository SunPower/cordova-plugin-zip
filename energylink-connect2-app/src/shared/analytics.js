import appVersion from '../macros/appVersion.macro'
import { capitalizeWord, getUserProfile } from 'shared/analyticsUtils'
import { MIXPANEL_EVENT_QUEUED } from 'state/actions/analytics'
import { getAppleDeviceFamily } from 'shared/appleDevicesTable'
import { prop } from 'ramda'

const getAppAndDeviceProperties = () => {
  const { device } = window
  const { model, version, platform, manufacturer } = device

  let deviceModel
  let deviceBrowserVersion

  if (platform === 'iOS') {
    deviceModel = getAppleDeviceFamily(model)
    deviceBrowserVersion = version
  } else {
    deviceModel = `${manufacturer}, ${model}`
  }

  return {
    'App Build': appVersion(),
    'App Flavor': process.env.REACT_APP_FLAVOR,
    'OS Version': version,
    ...(deviceBrowserVersion ? { $browser_version: deviceBrowserVersion } : {}),
    $device: deviceModel
  }
}

export const loggedIn = user => {
  const { mixpanel } = window
  const {
    uniqueId: userId,
    firstName,
    lastName,
    email,
    dealerName,
    recordType: dealerType
  } = getUserProfile(user)
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.identify(userId)
  mixpanel.people.set(appAndDeviceProperties)
  mixpanel.people.set({
    $first_name: firstName,
    $last_name: lastName,
    $email: email,
    'Dealer Name': dealerName,
    'Dealer Type': capitalizeWord(dealerType)
  })
  mixpanel.track('Login', { Success: true })

  return MIXPANEL_EVENT_QUEUED('Login - Success')
}

export const loginFailed = () => {
  const { mixpanel } = window
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.track('Login', { Success: false })

  return MIXPANEL_EVENT_QUEUED('Login - Failed')
}

export const scanPVS = ({ pvsSN, entryMethod }, event) => {
  const { mixpanel } = window

  mixpanel.register({ 'PVS SN': pvsSN })
  mixpanel.track('Scan PVS Tag', {
    'Entry Method': entryMethod
  })

  return MIXPANEL_EVENT_QUEUED(event)
}

export const siteNotFound = ({
  uniqueId: userId,
  firstName,
  lastName,
  email,
  dealerName,
  recordType: dealerType
}) => {
  const { mixpanel } = window
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.identify(userId)
  mixpanel.people.set(appAndDeviceProperties)
  mixpanel.people.set({
    $first_name: firstName,
    $last_name: lastName,
    $email: email,
    'Dealer Name': dealerName,
    'Dealer Type': capitalizeWord(dealerType)
  })
  mixpanel.track('Find Site', { Found: false })

  return MIXPANEL_EVENT_QUEUED('Find Site - site not found')
}

export const siteFound = (
  {
    uniqueId: userId,
    firstName,
    lastName,
    email,
    dealerName,
    recordType: dealerType
  },
  { city, st_id, postalCode, siteKey, commissioned }
) => {
  const { mixpanel } = window
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.identify(userId)
  mixpanel.people.set(appAndDeviceProperties)
  mixpanel.people.set({
    $first_name: firstName,
    $last_name: lastName,
    $email: email,
    'Dealer Name': dealerName,
    'Dealer Type': capitalizeWord(dealerType)
  })
  mixpanel.track('Find Site', {
    Found: true,
    ...{
      $city: city,
      State: st_id,
      'Zip Code': postalCode,
      'Site ID': siteKey,
      Commissioned: commissioned
    }
  })

  return MIXPANEL_EVENT_QUEUED('Find Site - site found')
}

export const saveConfiguration = config => {
  const { mixpanel } = window
  mixpanel.track('Configure', config)
  return MIXPANEL_EVENT_QUEUED('Configure - submit site config')
}

export const saveInventory = (inventory, hasEss) => {
  const { mixpanel } = window

  mixpanel.register({
    'Equinox Storage Site': hasEss
  })

  mixpanel.track('Inventory', {
    'AC Module Count': prop('AC_MODULES', inventory),
    'ESS System Size': hasEss ? prop('ESS', inventory) : 'None',
    'Equinox Storage Site': hasEss
  })
  return MIXPANEL_EVENT_QUEUED('Inventory')
}

export const deviceResume = pageResume => {
  const { mixpanel } = window
  mixpanel.register({
    Page: pageResume
  })
  return MIXPANEL_EVENT_QUEUED('User opens app - Register page resume')
}

/**
 *
 * @param {string}[interfaceType] - The type of connection the user has. either can be  PVS, Cell, or Customer WiFi
 * @return {Action<null, null>}
 */
export const internetConnection = (interfaceType = '') => {
  const { mixpanel } = window
  mixpanel.register({
    'Internet Connection': interfaceType
  })
  return MIXPANEL_EVENT_QUEUED('User opens app - Register page resume')
}

export const firmwareUpdateEvent = data => {
  const { mixpanel } = window
  mixpanel.track('Firmware Update', {
    'Current FW Version': prop('fromFWVersion', data),
    'Current BN Version': prop('fromBuildNumber', data),
    'New FW Version': prop('toFWVersion', data),
    'New BN Version': prop('toBuildNumber', data),
    Success: prop('success', data)
  })
  return MIXPANEL_EVENT_QUEUED('Firmware update - Update successful')
}

export const timeMixPanelEvent = name => {
  if (!name) throw new Error('The name parameter is required')
  const { mixpanel } = window
  mixpanel.time_event(name)
}

export const commissionSite = ({ duration, timeConfiguring }) => {
  const { mixpanel } = window
  mixpanel.track('Commission Site', {
    $duration: duration,
    'Time Configuring': timeConfiguring
  })
  return MIXPANEL_EVENT_QUEUED(
    'Commission Site - first successful site configuration'
  )
}
