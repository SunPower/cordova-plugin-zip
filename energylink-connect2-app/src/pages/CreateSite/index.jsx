import TextField from '@sunpower/textfield'
import SearchField from 'components/SearchField'
import { path } from 'ramda'
import React, { useEffect, useState } from 'react'
import { useField, useForm } from 'react-final-form-hooks'
import PlacesAutocomplete from 'react-places-autocomplete'
import { useDispatch } from 'react-redux'
import { Link } from 'react-router-dom'
import paths from 'routes/paths'
import { useI18n } from 'shared/i18n'
import { geocodeByAddress, getGeocodeData } from 'shared/utils'

function useGoogleMaps() {
  const [hasGoogleMaps, setGoogleMaps] = useState(!!window.google)
  useEffect(() => {
    setGoogleMaps(!!window.google)
  }, [])
  return hasGoogleMaps
}

function CreateSite() {
  const hasGoogleMaps = useGoogleMaps()
  const [initialValues, setInitialValues] = useState({
    siteName: '',
    apt: '',
    city: '',
    state: '',
    postalCode: '',
    address: ''
  })
  const t = useI18n()
  const dispatch = useDispatch()

  const { form, handleSubmit } = useForm({
    onSubmit: onSubmit(dispatch),
    initialValues,
    validate: values => null
  })
  const siteName = useField('siteName', form)
  const apt = useField('apt', form)
  const city = useField('city', form)
  const state = useField('state', form)
  const postalCode = useField('postalCode', form)
  useField('lat', form)
  useField('lng', form)
  /**
   * All this handle is just to populate all the fields
   * @param address
   */
  const handleSelect = async address => {
    const geocodeData = await geocodeByAddress(address.value)
    const { parsedData, lat, lng } = getGeocodeData(geocodeData)
    setInitialValues({
      siteName: siteName.input.value,
      city: parsedData.locality || parsedData.political,
      postalCode: parsedData.postal_code,
      state: parsedData.administrative_area_level_1,
      lat,
      lng,
      address: address.value
    })
  }
  if (!hasGoogleMaps)
    return (
      <section className="is-flex tile is-vertical section pt-0 fill-parent">
        <h1 className="has-text-centered is-uppercase has-text-weight-bold  pb-20">
          {t('WAITING_GOOGLE_MAPS')}
        </h1>
      </section>
    )
  return (
    <PlacesAutocomplete
      value={initialValues.address}
      onSelect={handleSelect}
      onChange={newAddress =>
        setInitialValues({
          ...initialValues,
          address: newAddress,
          siteName: siteName.input.value,
          city: city.input.locality,
          postalCode: postalCode.input.postal_code,
          state: state.input.administrative_area_level_1
        })
      }
    >
      {({ getInputProps, suggestions }) => (
        <section className="is-flex tile is-vertical section pt-0 fill-parent">
          <h1 className="has-text-centered is-uppercase has-text-weight-bold  pb-20">
            {t('CREATE_SITE')}
          </h1>

          <form onSubmit={handleSubmit} className="mt-20 pt-20">
            <article className="mb-15">
              <label htmlFor="siteName" className="has-text-white">
                {t('SITE_NAME')}
              </label>
              <TextField
                input={siteName.input}
                meta={siteName.meta}
                type="text"
                autoComplete="site_name"
                className="mt-5"
              />
            </article>

            <article className="mb-15">
              <label htmlFor="address" className="has-text-white">
                {t('ADDRESS')}
              </label>
              <SearchField
                onSearch={async newAddress => {
                  getInputProps().onChange({ target: { value: newAddress } })
                  return suggestions
                    .filter(elem => {
                      const country = path(
                        ['terms', elem.terms.length - 1, 'value'],
                        elem
                      )
                      return (
                        //English based language phone
                        country === 'USA' ||
                        country === 'Mexico' ||
                        country === 'Canada' ||
                        //spanish based language phone
                        country === 'EE. UU.' ||
                        country === 'México' ||
                        country === 'Canadá' ||
                        // french based language phone
                        country === 'États-Unis' ||
                        country === 'Mexique'
                      )
                    })
                    .map(elem => ({
                      value: elem.description,
                      label: elem.description
                    }))
                }} //load options
                onSelect={handleSelect} //onchange
                className="mt-10"
              />
            </article>

            <div className="is-flex level">
              <article className="mb-15 mr-5">
                <label htmlFor="apt" className="has-text-white">
                  {t('APT')}
                </label>
                <TextField
                  input={apt.input}
                  meta={apt.meta}
                  type="text"
                  autoComplete="apartment"
                  className="mt-5"
                />
              </article>

              <article className="mb-15 ml-5">
                <label htmlFor="city" className="has-text-white">
                  {t('CITY')}
                </label>
                <TextField
                  input={city.input}
                  meta={city.meta}
                  type="text"
                  autoComplete="city"
                  className="mt-5"
                />
              </article>
            </div>

            <div className="is-flex level">
              <article className="mb-15 mr-5">
                <label htmlFor="state" className="has-text-white">
                  {t('STATE')}
                </label>
                <TextField
                  input={state.input}
                  meta={state.meta}
                  type="text"
                  autoComplete="state"
                  className="mt-5"
                />
              </article>

              <article className="mb-15 ml-5">
                <label htmlFor="postalCode" className="has-text-white">
                  {t('POSTAL_CODE')}
                </label>
                <TextField
                  input={postalCode.input}
                  meta={postalCode.meta}
                  type="text"
                  autoComplete="postal code"
                  className="mt-5"
                />
              </article>
            </div>

            <div className="is-flex file level space-around section pt-0">
              <button
                className="button is-primary is-uppercase mb-15"
                type="submit"
              >
                {t('CREATE')}
              </button>

              <Link to={paths.PROTECTED.ROOT.path}>{t('CANCEL')}</Link>
            </div>
          </form>
        </section>
      )}
    </PlacesAutocomplete>
  )
}

export default CreateSite

function onSubmit(dispatch) {
  return console.info
}
