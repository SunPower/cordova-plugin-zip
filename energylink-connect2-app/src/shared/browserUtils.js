import { createExternalLinkHandler } from 'shared/routing'

export function openUrl(url) {
  window.SafariViewController.isAvailable(function(available) {
    if (available) {
      window.SafariViewController.show({
        url: url,
        hidden: false, // default false. You can use this to load cookies etc in the background (see issue #1 for details).
        animated: false, // default true, note that 'hide' will reuse this preference (the 'Done' button will always animate though)
        transition: 'curl', // (this only works in iOS 9.1/9.2 and lower) unless animated is false you can choose from: curl, flip, fade, slide (default)
        enterReaderModeIfAvailable: false, // default false
        tintColor: '#ffffff', // default is ios blue
        barColor: '#15202e', // on iOS 10+ you can change the background color as well
        controlTintColor: '#15202e' // on iOS 10+ you can override the default tintColor
      })
    } else {
      // potentially powered by InAppBrowser because that (currently) clobbers window.open
      createExternalLinkHandler(url)()
    }
  })
}

export function dismissBrowser() {
  window.SafariViewController.hide()
}
