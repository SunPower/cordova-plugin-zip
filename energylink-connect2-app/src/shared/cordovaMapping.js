import { fileExists } from 'shared/fileSystem'

export const getMd5FromFile = (filePath = '') =>
  new Promise((resolve, reject) => {
    fileExists(filePath).then(fileEntry => {
      if (!fileEntry) throw Error("The file doesn't exist")
      window.md5chksum.file(fileEntry, resolve, reject)
    })
  })

export const checkMD5 = async (filePath = '', expectedMd5 = '') => {
  const md5 = await getMd5FromFile(filePath)
  if (expectedMd5 === md5) return true
  throw new Error('MD5 are not the same')
}

export const cdvfileToNativePath = (filePath = '') =>
  new Promise((resolve, reject) => {
    fileExists(filePath)
      .then(fileEntry => {
        if (!fileEntry) reject("The file doesn't exists")
        resolve(fileEntry.toURL())
      })
      .catch(reject)
  })

export const getMD5FromFile = (filePath = '') => {
  return new Promise((resolve, reject) => {
    cdvfileToNativePath(filePath)
      .then(nativePath => {
        window.FileHash.sha256(
          nativePath,
          ({ result }) => resolve(result),
          reject
        )
      })
      .catch(reject)
  })
}
