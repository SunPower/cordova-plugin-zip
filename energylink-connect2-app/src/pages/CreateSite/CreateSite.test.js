import React from 'react'
import * as reactRedux from 'react-redux'

import CreateSite from '.'

import * as i18n from 'shared/i18n'

describe('CreateSite component', () => {
  describe('initial state', () => {
    let dispatchMock
    let initialState = {
      site: {
        isFetching: false,
        isSaving: false,
        sites: [],
        site: null,
        error: null,
        saveError: null,
        mapViewSrc: false
      }
    }

    beforeEach(() => {
      dispatchMock = jest.fn()
      jest
        .spyOn(reactRedux, 'useDispatch')
        .mockImplementation(() => dispatchMock)
      jest
        .spyOn(i18n, 'useI18n')
        .mockImplementation(path => (key, ...params) =>
          `${key.toUpperCase()} ${params.join('_')}`.trim()
        )
    })

    test('render correctly', () => {
      const component = mountWithProvider(<CreateSite />)(initialState)
      expect(component).toMatchSnapshot()
    })
  })

  describe('with error', () => {
    let dispatchMock
    let initialState = {
      site: {
        isFetching: false,
        isSaving: false,
        sites: [],
        site: null,
        error: true,
        saveError: null,
        mapViewSrc: false
      }
    }

    beforeEach(() => {
      dispatchMock = jest.fn()
      jest
        .spyOn(reactRedux, 'useDispatch')
        .mockImplementation(() => dispatchMock)
      jest
        .spyOn(i18n, 'useI18n')
        .mockImplementation(path => (key, ...params) =>
          `${key.toUpperCase()} ${params.join('_')}`.trim()
        )
    })

    test('render correctly', () => {
      const component = mountWithProvider(<CreateSite />)(initialState)
      expect(component).toMatchSnapshot()
    })
  })
})
