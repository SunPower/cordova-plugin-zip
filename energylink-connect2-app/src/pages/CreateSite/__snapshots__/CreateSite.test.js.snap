// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateSite component initial state render correctly 1`] = `
Object {
  "component": <BrowserRouter
    keyLength={0}
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Provider
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
      >
        <CreateSite>
          <section
            className="is-flex tile is-vertical section pt-0 fill-parent"
          >
            <h1
              className="has-text-centered is-uppercase has-text-weight-bold  pb-20"
            >
              WAITING_GOOGLE_MAPS
            </h1>
          </section>
        </CreateSite>
      </Provider>
    </Router>
  </BrowserRouter>,
  "store": Object {
    "clearActions": [Function],
    "dispatch": [Function],
    "getActions": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
}
`;

exports[`CreateSite component with error render correctly 1`] = `
Object {
  "component": <BrowserRouter
    keyLength={0}
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Provider
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
      >
        <CreateSite>
          <section
            className="is-flex tile is-vertical section pt-0 fill-parent"
          >
            <h1
              className="has-text-centered is-uppercase has-text-weight-bold  pb-20"
            >
              WAITING_GOOGLE_MAPS
            </h1>
          </section>
        </CreateSite>
      </Provider>
    </Router>
  </BrowserRouter>,
  "store": Object {
    "clearActions": [Function],
    "dispatch": [Function],
    "getActions": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
}
`;
